Clarify the 3 primitives used for defining policy targets (#23066)
kubernetes/website: content/en/docs/concepts/services-networking/network-policies.md

---

<!-- overview -->

> A Kubernetes network policy is a specification of how groups of {{< glossary_tooltip text="pods" term_id="pod">}} are allowed to communicate with endpoints.  The endpoints which a Pod's NetworkPolicy will allow traffic to, or from, is be defined by combining one or more of the following constructs:

Reviewer: I worry that this might get confused with Endpoints / EndpointSlice. Is there a word we can use in place of endpoints?

Author: updated. with a few other deets, good point

1. pods (based on labels)
2. namespaces (based on labels)
3. IP blocks (based on CIDR definitions)

> When defining a pod or namespace based NetworkPolicy, we use {{< glossary_tooltip text="labels" term_id="label">}} to select pods and define rules which specify what traffic is allowed to the selected pods.

Reviewer: Is this more accurate:
```suggestion
When defining a pod- or namespace- based NetworkPolicy, you use a {{< glossary_tooltip text="selector" term_id="selector">}} to specify what traffic is allowed to and from the Pod(s) that match the selector.
```
?

---

<!-- overview -->

> If you want to control traffic flow at the IP address or port level (OSI layer 3 or 4), then you might consider using Kubernetes NetworkPolicies for particular applications in your cluster.  NetworkPolicies are an application-centric construct which allow you to specify how a {{< glossary_tooltip text="Pods" term_id="pod">}} is allowed to communicate with various network "entities" (we use the word "entity" here to avoid overloading the more common terms such as "endpoints" and "services", which have specific Kubernetes connotations) over the network. 

Reviewer 2: @jayunit100 Jay, Pod here should be singular, right? PTAL at this: https://deploy-preview-23066--kubernetes-io-master-staging.netlify.app/docs/concepts/services-networking/network-policies/

Or maybe, as Network Policy is defined by Label, turn the whole phrase into plural: how some Pods are allowed to communicate with [...]

> The entities which a Pod can communicate with must be "selected" based on a combination of the following 3 identifiers.
> 
> 1. Other pods are allowed (exception: a pod cannot block access to itself)
> 2. namespaces which are allowed 

Reviewer: Is this also accurate:
```suggestion
The entities that a Pod can communicate with are identified through a combination of the following 3 identifiers:

1. Other pods that are allowed (exception: a pod cannot block access to itself)
2. Namespaces that are allowed 
```

> 3. IP blocks (exception: a pod's host node is always allowed, regardless of it's IP) 

Reviewer: “it's” should be “its”, or:
```suggestion
3. IP blocks (exception: traffic to and from the node where a Pod is running is always allowed, regardless of the IP address of the Pod or the node)
```

When defining a pod- or namespace- based NetworkPolicy, you use a {{< glossary_tooltip text="selector" term_id="selector">}} to specify what traffic is allowed to and from the Pod(s) that match the selector.

> Meanwhile, when IP based NetworkPolicys are created, we define policies based on IP blocks (CIDR ranges).

Reviewer 2: s/NetworkPolicys/NetworkPolicies/

You must be using a {{< glossary_tooltip text="CNI" term_id="cni" >}} plugin that supports SCTP protocol NetworkPolicies.
{{< /note >}}

# What you CAN'T do with network policy's (at least, not yet)

> As of Kubernetes 1.20, the following functionality does not exist in the NetworkPolicy API, but you might be able to implement workarounds using Operating System components (such as SELinux, OpenVSwitch, IPTables, and so on) or Layer 7 technologies (Ingress controllers, Service Mesh implementations) or admission controllers.  In case you are new to network security in Kubrnetes, its worth noting that the following User Stories cannot (yet) be implemented using the NetworkPolicy API.  Some (but not all) of these user stories are actively being discussed for future releases of the NetworkPolicy API.

Reviewer 2: s/Kubrnetes/Kubernetes

- Forcing internal cluster traffic to go through a common gateway (this might be best served with a service mesh or other proxy).
- Anything TLS related (use a service mesh or ingress controller for this).
- Node specific policies (you can use CIDR notation for these, but you cannot target nodes by their Kubernetes identities specifically).

> - Targeting of namespaces or services by name (you can, however, target pods or namespaces by their{{< glossary_tooltip text="labels" term_id="label" >}}, which is often an viable workaround).

Reviewer 2: "a viable workaround"

- Creation or management of "Policy requests" that are fulfilled by a third party.
- Default policies which are applied to all namespaces or pods (there are some third party Kubernetes distributions and projects which can do this).
- Advanced policy querying and reachability tooling.
- The ability to target ranges of Ports in a single policy declaration.
- The ability to log network security events (for example connections that are blocked or accepted).

> - The ability to explicitly deny policies (currently the model for NetworkPolicies are deny by default, with only the ability to add allow rules).

Reviewer 2: I've a specific question on this (sorry, didn't tested yet): when we define a ingress policy, it's assumed that all the ingress traffic will be blocked BUT that specified? Is this per namespace or per Pod selector? 

As far as I remember, the behavior from ingress policy is different from egress policy (I remember a discussion about this with Casey or Chris Luciano), so maybe in this or in a later commit this could be explored (and maybe drawed as a flow, don't know) so that users would know that ingress and egress have different default behavior :)

Author: I thought egress deny all is the same except usually people open up port 53 so that at least dns can work, so that later once real egress allows are layered on the deny the policies work intuitively .

Lmk if u know if anything different I'm happy to test for you.

Reviewer 2: Yeap, the 'default deny' example. My point here was just that as far as I remember the behavior of ingress and egress are different when a first 'specific' rule was created, but in this case nevermind 

I've left some other reviews above
